runtime/lf_skiplist.c

```diff
@@ -444,55 +444,52 @@ int caml_lf_skiplist_remove(struct lf_skiplist *sk, uintnat key) {
-       successor nodes at each level. */
-    int found = skiplist_find(sk, key, preds, succs);
+  /* As with insert. If the node doesn't exist, we don't need to do anything.
+     While we're checking for it we populate the predecessor nodes and successor
+     nodes at each level. */
+  int found = skiplist_find(sk, key, preds, succs);
 
-    if (!found) {
-      return 0;
-    } else {
-      /* When the node exists in the skiplist, then succs[0] must point to it.
-         Note: this isn't the case for levels > 0. */
-      struct lf_skipcell *to_remove = succs[0];
-      for (int level = to_remove->top_level; level >= 1; level--) {
-        /* We mark each of the forward pointers at every level the node is
-           present at. We may be raced by another thread deleting the same node
-           and by threads inserting new nodes directly after the node we are
-           removing, so we need to retry the CAS in a loop to deal with the
-           latter. */
+  if (!found) {
+    return 0;
+  } else {
+    /* When the node exists in the skiplist, then succs[0] must point to it.
+       Note: this isn't the case for levels > 0. */
+    struct lf_skipcell *to_remove = succs[0];
+    for (int level = to_remove->top_level; level >= 1; level--) {
+      /* We mark each of the forward pointers at every level the node is present
+         at. We may be raced by another thread deleting the same node and by
+         threads inserting new nodes directly after the node we are removing,
+         so we need to retry the CAS in a loop to deal with the latter. */
+      LF_SK_EXTRACT(to_remove->forward[level], marked, succ);
+
+      while (!marked) {
+        atomic_compare_exchange_strong(&to_remove->forward[level], &succ,
+                                       LF_SK_MARKED(succ));
         LF_SK_EXTRACT(to_remove->forward[level], marked, succ);
-
-        while (!marked) {
-          atomic_compare_exchange_strong(&to_remove->forward[level], &succ,
-                                         LF_SK_MARKED(succ));
-          LF_SK_EXTRACT(to_remove->forward[level], marked, succ);
-        }
       }
+    }
 
-      /* The bottom layer is what ultimately determines whether the node is
-         present in the skiplist or not. We try to remove it and if we succeed
-         then indicate so to the caller. If not then another thread raced us an
-         won. */
-      LF_SK_EXTRACT(to_remove->forward[0], marked, succ);
-      while (1) {
-        int mark_success = atomic_compare_exchange_strong(
-            &to_remove->forward[0], &succ, LF_SK_MARKED(succ));
-
-        LF_SK_EXTRACT(to_remove->forward[0], marked, succ);
+    /* The bottom layer is what ultimately determines whether the node is
+       present in the skiplist or not. We try to remove it and if we succeed
+       then indicate so to the caller. If not then another thread raced us an
+       won. */
+    LF_SK_EXTRACT(to_remove->forward[0], marked, succ);
+    while (1) {
```

Comment:
We do need a loop here. This loop body has three possible exits:

1. We successfully marked the bottom-level forward reference, which means that we logically removed the element. The skiplist_find call will physically remove it.
2. We failed to acquire-mark the bottom-level forward reference, because another thread did what we wanted to do, so the element is certainly logically removed and either has been physically removed or will be soon.
3. We failed to acquire-mark the bottom-level forward reference, because another thread changed the reference to another unmarked value. Imagine adding a new element subsequent to the to_remove element. We need to mark the new reference, so we loop.